### ISSUE TITLE
Selecting socket for outgoing requests (src_any)

### ISSUE THREAD

**Issue by flicker581** [FreeRADIUS/freeradius-server#1148]
I hardly believe that I am writing this, after all years, but nevertheless... It seemes to me that this is a bug. Please note that we use FreeRADIUS 2.2 here, but it seemes to apply to current versions as well.

In a complex configuration, when there are different proxy destinations, some may define src_ipaddr, and some may use default (which is equivalent to 0.0.0.0, or INADDR_ANY). When using INADDR_ANY, the expected behaviour is to let networking subsystem in kernel to decide which source address to use, based on routing and outgoing interface. To do this, FreeRADIUS should use a socket with source address INADDR_ANY. 

But now, when sending a request with src_any, FreeRADIUS treats it as "don't care" and takes first available socket with any address. As a result, outgoing packets are sometimes sent with unexpected source addresses. In some configurations, they may end up with source of 127.0.0.1, which is really scary.

In our case, we proxy some requests to Home_A with specific src_ipaddr IP_2, which is a secondary address (alias) of outgoing interface. We also proxy to Home_B without specific src_ipaddr, and it is expected to use primary address of the outgoing interface (IP_1). (We do not want to specify src_ipaddr for Home_B because it will make configs in our cluster slightly different, whereas now they are identical). In result, the server is proxying to Home_B from both IP_1 and IP_2.

The same problem was described here:

http://lists.freeradius.org/pipermail/freeradius-users/2011-October/056573.html

As I understand (I can be wrong, although it seemes pretty obvious), this code in packet.c is responsible for socket selection:

		/*
		 *	We're sourcing from *, and they asked for a
		 *	specific source address: ignore it.
		 */
		if (ps->src_any && !src_any) continue;

		/*
		 *	We're sourcing from a specific IP, and they
		 *	asked for a source IP that isn't us: ignore
		 *	it.
		 */
		if (!ps->src_any && !src_any &&
		    (fr_ipaddr_cmp(&request->src_ipaddr,
				   &ps->src_ipaddr) != 0)) continue;

And my straightforward, but unfortunately untested, patch is to add:

               /*
                *      And vice versa, we have a specific address,
                *      but they want to use INADDR_ANY. Ignore.
                */
               if (!ps->src_any && src_any) continue;

I am really sorry to present this unverified patch. Hope the issue is obvious enough.

**Comment by alandekok** [comment]
Unfortunately, that patch won't work.  If people set the IP address for the proxy socket, then the server will be unable to proxy *any* packets.

The short fix is to ensure that we don't pick source 127/8 unless the destination is also 127/8.

The real fix is to set the source IP address in the home_server configuration.  That is designed to avoid precisely this situation.

**Comment by flicker581** [comment]
Alan, please, don't hurry. 127/8 is just a scary example, this does not need a particular fix. The same problem comes back when a multihomed server has 192.168.1.1 (private) IP on one interface and a public address on the other interface. The public address can be even dynamic on ppp0. Then the server may end up sending from 192.168.1.1 to the Internet, which will not work.

Setting source IP in home_server is working workaround, but not always convenient. As if the address on ppp0 which is dynamic.

The point is to allow user send from INADDR_ANY, if he wants so.

**Comment by alandekok** [comment]
Please read my earlier message.  It is possible to configure the server so that *all* proxy listeners have an IP address.  If the server then tries to proxy a packet, and doesn't care about the source IP... your patch will make *all* proxying fail.

And your comment is wrong, too.  It is perfectly valid for a system to route packets from 192.168/16 to public IPs, even if it has a public IP on a second interface.  The first interface may be 192.168/16, *and* be the default route for packets.

That final sentence is key.  In the end, this is a routing problem.  The patch I put in for 127/8 is correct.  In general, no other IP address can route to 127/8, and no packets from 127/8 can be routed to any other IP address.

The *only* solution is to explicitly set the source IP for the home server in the configuration files.  Either that, or you need to submit a patch which makes FreeRADIUS snoop through the local machines routing table and list of interfaces, to determine which source socket can be used to send packets to a particular destination.

In short, you can't solve routing problems by poking the application.  Your patch is wrong, and will break the server for people who explicitly list outgoing proxy IP addresses.  Which is allowed.  And which they do.
